server/src/inmantals/server.py

```diff
@@ -75,22 +216,57 @@ def __init__(self, instream: BaseIOStream, outstream: BaseIOStream, address):
         self.diagnostics_cache: Optional[lsp_types.PublishDiagnosticsParams] = None
         self.supported_symbol_kinds: Optional[Set[lsp_types.SymbolKind]] = None
         # compiler_venv_path is only relevant for versions of core that require a compiler venv. It is ignored otherwise.
-        self.compiler_venv_path: Optional[str] = None
+        self.inmanta_settings: Dict[str, str] = {}
+
+    def add_setting(self, key: str, value: str) -> None:
+        """Register a default value for a setting."""
+        if value is not None:
+            self.inmanta_settings[key] = value
 
-    async def initialize(self, rootPath, rootUri, capabilities: Dict[str, object], **kwargs):  # noqa: N803
+    def get_setting(self, key, folder: Optional[Folder] = None, default: Optional[str] = None) -> Optional[str]:
+        """
+        Fetch a given setting in the provided folder. If the setting isn't found or no folder is provided, this will look
+        for the default value for this setting.
+        """
+        if folder:
+            folder_setting = folder.get_setting(key, default)
+            return folder_setting if folder_setting else self.get_default_setting(key, default)
+        return self.get_default_setting(key, default)
+
+    def get_default_setting(self, key: str, default: Optional[str] = None) -> Optional[str]:
+        """Reads the default value for a setting set by the initializationOptions argument in the initialize method."""
+        logger.info(f"key={key} self.inmanta_settings={self.inmanta_settings}")
+        return self.inmanta_settings.get(key, default)
+
+    def remove_folder(self, folder_uri: str) -> None:
+        logger.info(self._workspace_folders)
+        del self._workspace_folders[folder_uri]
+
+    async def initialize(
+        self, workspaceFolders: Sequence[object], capabilities: Dict[str, object], root_path=None, root_uri=None, **kwargs
+    ):  # noqa: N803
         logger.debug("Init: " + json.dumps(kwargs))
+        logger.info(f"workspaceFolders={workspaceFolders}")
+
+        if root_path:
+            logger.warning("The rootPath parameter has been deprecated in favour of the 'workspaceFolders' parameter.")
+        if root_uri:
+            logger.warning("The rootUri parameter has been deprecated in favour of the 'workspaceFolders' parameter.")
```

Comment:
This one raised an exception before, why not anymore?

Reply:
I'll read the docs you linked and see if that changes anything for me, thanks.

Reply:
Is this part something you cover in the other PR? "Since workspace folders can come and go dynamically the server also needs to register for workspace/didChangeWorkspaceFolders notifications."

Reply:
> Is this part something you cover in the other PR
Yes, I ended up using the rootUri after all in the other PR

Reply:
I mean the part I quoted about `workspace/didChangeWorkspaceFolders`. It's missing here but if it's part of the other PR that's all right. So is it?

Reply:
Yes it is handled in the other PR, although not directly at the language server level, since each ls is only responsible for one folder in the workspace, so in a sense they don't care about the changes to the workspace.
It is handled in the extension, namely keeping track of each top level opened folder and its associated ls, and listening to changes to the workspace and starting/stopping the right ls when a folder is added/removed to the workspace.

Reply:
I see. That probably suffices. In any case, I'll consider that part of the other PR's review.